gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractCheckNativeDistributionRuntime.kt

```diff
@@ -36,41 +36,50 @@ abstract class AbstractCheckNativeDistributionRuntime : AbstractComposeDesktopTa
     @get:LocalState
     val workingDir: Provider<Directory> = taskDir.map { it.dir("localState") }
 
-    private val javaExec: File
-        get() = getTool("java")
+    private val jdkHomeFile: File
+        get() = File(jdkHome.orNull ?: error("Missing jdkHome value"))
 
-    private val javacExec: File
-        get() = getTool("javac")
+    private fun File.getJdkTool(toolName: String): File =
+        resolve("bin/${executableName(toolName)}").apply {
+            check(exists()) {
+                jdkDistributionProbingError("'$toolName' is missing")
+            }
+        }
 
-    private fun getTool(toolName: String): File {
-        val javaHomeBin = File(javaHome.get()).resolve("bin")
-        val tool = javaHomeBin.resolve(executableName(toolName))
-        check(tool.exists()) { "Could not find $tool at: ${tool.absolutePath}}" }
-        return tool
+    private fun jdkDistributionProbingError(errorMessage: String): Nothing {
+        val fullErrorMessage = buildString {
+            appendLine("Failed to check JDK distribution: $errorMessage")
+            appendLine("JDK distribution path: ${jdkHomeFile.absolutePath}")
+        }
+        error(fullErrorMessage)
     }
 
     @TaskAction
     fun run() {
         taskDir.ioFile.mkdirs()
 
-        val javaRuntimeVersion = try {
-            getJavaRuntimeVersionUnsafe()?.toIntOrNull() ?: -1
-        } catch (e: Exception) {
-            throw IllegalStateException(
-                "Could not infer Java runtime version for Java home directory: ${javaHome.get()}", e
-            )
-        }
+        val jdkHome = jdkHomeFile
+        val javaExecutable = jdkHome.getJdkTool("java")
+        val javacExecutable = jdkHome.getJdkTool("javac")
 
-        check(javaRuntimeVersion >= MIN_JAVA_RUNTIME_VERSION) {
-            """|Packaging native distributions requires JDK runtime version >= $MIN_JAVA_RUNTIME_VERSION
-               |Actual version: '${javaRuntimeVersion ?: "<unknown>"}'
-               |Java home: ${javaHome.get()}
-            """.trimMargin()
+        val jvmRuntimeVersionString = getJavaRuntimeVersion(
+            javaExecutable = javaExecutable,
+            javacExecutable = javacExecutable
+        )
+
+        val jvmRuntimeVersion = jvmRuntimeVersionString?.toIntOrNull()
```

Comment:
Applying it to `System.getProperty("java.version")` doesn't seem right because it might contain something like `11.0.4`